src/algebra/module.lean

```diff
@@ -99,17 +97,68 @@ lemma finset.sum_smul {f : ι → R} {s : finset ι} {x : M} :
   s.sum f • x = s.sum (λ r, (f r) • x) :=
 ((smul_add_hom R M).flip x).map_sum f s
 
-end semimodule
+end add_comm_monoid
 
-section prio
-set_option default_priority 100 -- see Note [default priority]
-/-- A module is a generalization of vector spaces to a scalar ring.
-  It consists of a scalar ring `R` and an additive group of "vectors" `M`,
-  connected by a "scalar multiplication" operation `r • x : M`
-  (where `r : R` and `x : M`) with some natural associativity and
-  distributivity axioms similar to those on a ring. -/
-class module (R : Type u) (M : Type v) [ring R] [add_comm_group M] extends semimodule R M
-end prio
+section add_comm_group
+
+variables (R M) [semiring R] [add_comm_group M]
+
+/-- A structure containing most informations as in a semimodule, except the fields `zero_smul`
+and `smul_zero`. As these fields can be deduced from the other ones when `M` is an `add_comm_group`,
+this provides a way to construct a semimodule structure by checking less properties, in
+`semimodule.of_core`. -/
+@[nolint has_inhabited_instance]
+structure semimodule.core extends has_scalar R M :=
+(smul_add : ∀(r : R) (x y : M), r • (x + y) = r • x + r • y)
+(add_smul : ∀(r s : R) (x : M), (r + s) • x = r • x + s • x)
+(mul_smul : ∀(r s : R) (x : M), (r * s) • x = r • s • x)
+(one_smul : ∀x : M, (1 : R) • x = x)
+
+variables {R M}
+
+/-- Define `semimodule` without proving `zero_smul` and `smul_zero` by using an auxiliary
+structure `semimodule.core`, when the underlying space is an `add_comm_group`. -/
+def semimodule.of_core (H : semimodule.core R M) : semimodule R M :=
+by letI := H.to_has_scalar; exact
+{ zero_smul := λ x,
+    have (0 : R) • x + (0 : R) • x = (0 : R) • x + 0, by rw ← H.add_smul; simp,
+    add_left_cancel this,
+  smul_zero := λ r,
+    have r • (0:M) + r • 0 = r • 0 + 0, by rw ← H.smul_add; simp,
+    add_left_cancel this,
```

Comment:
I have these two proofs in an old local `linear-semimodule` branch:
```
{ zero_smul := λ x, (add_monoid_hom.mk' (λ r : R, r • x) (λ r s, H.add_smul r s x)).map_zero,
  smul_zero := λ r, (add_monoid_hom.mk' ((•) r) (H.smul_add r)).map_zero,
```

Reply:
I started this branch a long time ago, so maybe this was the old proof and when I merged master it remained there. Anyway, replaced with the new proof.